### PR TITLE
Workaround cli product creation issue

### PR DESCRIPTION
**Pull request by sthirugn**
Sometimes cli product create errors for no reason when there is multiple product creation requests at the sametime although the product entities are created.  This workaround will attempt to wait for 5 seconds and query the product again to make sure it is actually created.  If it is not found, the setUpClass will fail and stop.  This issue will have to be looked at detail later to make sure that this is not a product issue.

### CODE REVIEW

**Comment by sthirugn** [comment]
```sh
py.test -v tests/foreman/cli/test_contentviewfilter.py
===================================================== test session starts =====================================================
platform linux2 -- Python 2.7.11, pytest-2.9.1, py-1.4.31, pluggy-0.3.1 -- /home/suresh/hacking/Envs/robottelo/bin/python
cachedir: .cache
rootdir: /home/suresh/hacking/robottelo, inifile: 
collected 29 items 

tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_with_invalid_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_with_invalid_repo_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_with_same_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_without_cv PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_without_type PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_delete_by_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_inclusion PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_invalid_repo_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_non_existent_repo_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_same_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_by_cv_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_by_org_label PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_by_org_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_content_type_by_cv_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_description_by_cv_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_name_by_cv_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_original_pkgs PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_repo_by_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_repo_by_name <- robottelo/decorators.py FAILED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_repos_yum_and_docker PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_delete_by_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_delete_by_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_delete_by_org_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_inclusion PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_repo_with_different_type PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_repo_with_same_type PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positivec_create_with_inclusion_by_cv_id PASSED
```

File: tests/foreman/cli/test_contentviewfilter.py [hunk 1]
```diff
@@ -25,7 +28,26 @@ def setUpClass(cls):
...
         super(ContentViewFilterTestCase, cls).setUpClass()
         cls.org = make_org()
-        cls.product = make_product({u'organization-id': cls.org['id']})
+        product_name = gen_string('alpha')
+        try:
+            cls.product = make_product({
+                u'organization-id': cls.org['id'],
+                u'name': product_name,
+            })
+        except CLIFactoryError:
```
Comment (elyezer) [inline tests/foreman/cli/test_contentviewfilter.py]: Should not be better to workaround while some BZ is open?

**Comment by elyezer** [comment]
ACK, pending comment.

File: tests/foreman/cli/test_contentviewfilter.py (hunk 1 above)
Comment (sthirugn) [inline tests/foreman/cli/test_contentviewfilter.py]: there is no bz yet, since this issue is not reproducible anywhere except jenkins :/

**Comment by sthirugn** [comment]
thanks for @elyezer for irc conversation about this pr.  Now the code is updated and it looks simple and straightforward.  I will run tests and paste results here.

**Comment by sthirugn** [comment]
Removing Do not merge to get the review going.

**Comment by sthirugn** [comment]
```sh
py.test -v tests/foreman/cli/test_contentviewfilter.py
==================================================================== test session starts =====================================================================
platform linux2 -- Python 2.7.11, pytest-2.9.1, py-1.4.31, pluggy-0.3.1 -- /home/suresh/hacking/Envs/robottelo/bin/python
cachedir: .cache
rootdir: /home/suresh/hacking/robottelo, inifile: 
plugins: cov-2.2.1, xdist-1.14
collected 29 items 

tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_with_invalid_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_with_invalid_repo_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_with_same_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_without_cv PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_create_without_type PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_delete_by_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_inclusion PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_invalid_repo_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_non_existent_repo_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_negative_update_with_same_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_by_cv_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_by_org_label PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_by_org_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_content_type_by_cv_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_description_by_cv_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_name_by_cv_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_original_pkgs PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_repo_by_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_repo_by_name <- robottelo/decorators.py FAILED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_create_with_repos_yum_and_docker PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_delete_by_id PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_delete_by_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_delete_by_org_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_inclusion PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_name PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_repo_with_different_type PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positive_update_repo_with_same_type PASSED
tests/foreman/cli/test_contentviewfilter.py::ContentViewFilterTestCase::test_positivec_create_with_inclusion_by_cv_id PASSED

========================================================================== FAILURES ==========================================================================
______________________________________________ ContentViewFilterTestCase.test_positive_create_with_repo_by_name ______________________________________________
```

Note: The one failure above is related to a test which is skipped due to a bug. My local tests are not getting skipped with bugzilla status :/

**Comment by sthirugn** [comment]
```sh
py.test -v tests/foreman/cli/test_docker.py
==================================================================== test session starts =====================================================================
platform linux2 -- Python 2.7.11, pytest-2.9.1, py-1.4.31, pluggy-0.3.1 -- /home/suresh/hacking/Envs/robottelo/bin/python
cachedir: .cache
rootdir: /home/suresh/hacking/robottelo, inifile: 
plugins: cov-2.2.1, xdist-1.14
collected 51 items 

tests/foreman/cli/test_docker.py::DockerManifestTestCase::test_positive_read_docker_tags PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_create_repos_using_multiple_products <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_create_repos_using_same_product <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_create_with_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_delete_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_delete_random_repo_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_sync <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_update_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_update_upstream_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRepositoryTestCase::test_positive_update_url <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_add_docker_repo_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_add_docker_repo_by_id_to_ccv <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_add_docker_repos_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_add_docker_repos_by_id_to_ccv <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_add_synced_docker_repo_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_promote_multiple_with_docker_repo <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_promote_multiple_with_docker_repo_composite <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_promote_with_docker_repo <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_promote_with_docker_repo_composite <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_publish_multiple_with_docker_repo <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_publish_multiple_with_docker_repo_composite <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_publish_with_docker_repo <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContentViewTestCase::test_positive_publish_with_docker_repo_composite <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerActivationKeyTestCase::test_positive_add_docker_repo_ccv <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerActivationKeyTestCase::test_positive_add_docker_repo_cv <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerActivationKeyTestCase::test_positive_remove_docker_repo_ccv <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerActivationKeyTestCase::test_positive_remove_docker_repo_cv <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerClientTestCase::test_positive_pull_image <- ../Envs/robottelo/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/cli/test_docker.py::DockerClientTestCase::test_positive_upload_image <- ../Envs/robottelo/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/cli/test_docker.py::DockerComputeResourceTestCase::test_positive_create_external <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerComputeResourceTestCase::test_positive_create_internal <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerComputeResourceTestCase::test_positive_delete_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerComputeResourceTestCase::test_positive_list_containers_internal <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerComputeResourceTestCase::test_positive_update_internal <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContainersTestCase::test_positive_create_using_cv <- robottelo/decorators.py FAILED
tests/foreman/cli/test_docker.py::DockerContainersTestCase::test_positive_create_with_compresource <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContainersTestCase::test_positive_create_with_external_registry <- robottelo/decorators.py SKIPPED
tests/foreman/cli/test_docker.py::DockerContainersTestCase::test_positive_delete_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerContainersTestCase::test_positive_power_on_off <- robottelo/decorators.py FAILED
tests/foreman/cli/test_docker.py::DockerContainersTestCase::test_positive_read_container_log <- robottelo/decorators.py FAILED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_create_with_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_delete_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_delete_by_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_description_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_description_by_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_name_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_name_by_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_url_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_url_by_name <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_username_by_id <- robottelo/decorators.py PASSED
tests/foreman/cli/test_docker.py::DockerRegistryTestCase::test_positive_update_username_by_name <- robottelo/decorators.py PASSED
```
Note: The one failure above is related to a test which is skipped due to a bug. My local tests are not getting skipped with bugzilla status :/

**Comment by sthirugn** [comment]
This PR is ready for final review and merge.

**Comment by omaciel** [comment]
@sthirugn couldn't we just update the existing `make_product` method instead? Maybe update the code that attempts to fetch the entity being created instead (and add an optional `wait=False` argument) so that we don't have to add a new method to work around this?

**Comment by sthirugn** [comment]
@omaciel :) I did the exact same code as you explained when I submitted the PR.  @elyezer and I discussed later and decided not to touch `make_product` method.

**Comment by omaciel** [comment]
@elyezer can you share your thoughts about this?

**Comment by rplevka** [comment]
is there a BZ created for the issue? So we don't mask the issue. I'd prefer using BZ + `if bz_open()` combo to make sure we won't forget about it

**Comment by sthirugn** [comment]
@rplevka https://github.com/SatelliteQE/robottelo/pull/3426#discussion_r61799649

**Comment by elyezer** [comment]
@omaciel I suggested not to introduce a new argument to the make_product since it will differ from other factories. I was not sure if this is a special case enough to allow breaking that. Creating a specialized factory function looked a better approach.

**Comment by omaciel** [comment]
@elyezer I dunno... adding a new method that does basically the same thing doesn't strike me as being a better approach. Does anyone else in @SatelliteQE/quality-engineers want to chime in?

**Comment by elyezer** [comment]
To me the best approach would be create a BZ and just work around while it is open. Not change any of the factory signatures. But seems that these issue is very random and not allowed creating a bug.

**Comment by sthirugn** [comment]
@elyezer bz is created and the import currently changes automatically based on bz.

**Comment by elyezer** [comment]
@sthirugn missed that. Thanks for the clarification.

Seems that there are two options 1) use separate functions (current approach) or 2) do the logic inside the make_product.

Having that running at the import time is not a good thing to have though. That will make an HTTP request during the import time which can make deadlock when importing the module. Having side effects at import time is not good to have.

**Comment by sthirugn** [comment]
@elyezer is there an other way to workaround if conditional import is not an option?

**Comment by elyezer** [comment]
@sthirugn do the if else inside make_produt:

```python
def make_product(...):
    if bz_is_open():
         return make_product_wait
    ....
````

That way we avoid the import time sideeffect but that will apply for all cases.

**Comment by sthirugn** [comment]
Yes I thought about this and I commented earlier here too. So far I saw only two entities affected by this issue.  I don't want to use workaround for all entities yet.

**Comment by elyezer** [comment]
@sthirugn so do the inverse then, call make_product on make_product_wait and import straight make_product_wait and rename to make_product for the affected modules.

Something like:

```python
def make_product_wait(...):
    ...
    try:
        ...
    except ...:
        if not bz_bug_is_open(...):
            raise err
        # remaining of the except block
```

**Comment by sthirugn** [comment]
@elyezer I was looking for an easy option to switch between make_product() and make_product_wait() with minimal code changes - hence I chose conditional importing.  

But you are right, it might lead to other issues, I agree with your suggestion.

**Comment by elyezer** [comment]
@sthirugn yes it would be better to have that at the import time. That would be the best solution, but due to some Robottelo specifics we need to make sure http requests and other operation happens at execution time and not import time.

**Comment by sthirugn** [comment]
Conditional import removed and code updated.  I will let @elyezer and @omaciel deal with deciding if they are okay with the new wrapper `make_product_wait()` or patching `make_product()` but I really like to get this merged before next build if possible, please.

**Comment by elyezer** [comment]
ACK

**Comment by omaciel** [comment]
ACK

**Comment by sthirugn** [comment]
yay this workaround worked in latest run
screenshot from 2016-05-07 20-50-33
screenshot from 2016-05-07 20-50-09

**Comment by omaciel** [comment]
I fixed another issue with the content view filter test shown here

**Pull request merged**